runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/ParDoTranslatorBatch.java

```diff
@@ -83,61 +86,74 @@
       ClassTag.apply(Tuple2.class);
 
   @Override
-  public void translate(ParDo.MultiOutput<InputT, OutputT> transform, Context cxt)
-      throws IOException {
-    String stepName = cxt.getCurrentTransform().getFullName();
-
-    SparkCommonPipelineOptions opts = cxt.getOptions().as(SparkCommonPipelineOptions.class);
-    StorageLevel storageLevel = StorageLevel.fromString(opts.getStorageLevel());
+  public boolean canTranslate(ParDo.MultiOutput<InputT, OutputT> transform) {
+    DoFn<InputT, OutputT> doFn = transform.getFn();
+    DoFnSignature signature = DoFnSignatures.signatureForDoFn(doFn);
 
-    // Check for not supported advanced features
     // TODO: add support of Splittable DoFn
-    DoFn<InputT, OutputT> doFn = transform.getFn();
     checkState(
-        !DoFnSignatures.isSplittable(doFn),
+        !signature.processElement().isSplittable(),
         "Not expected to directly translate splittable DoFn, should have been overridden: %s",
         doFn);
 
     // TODO: add support of states and timers
     checkState(
-        !DoFnSignatures.isStateful(doFn), "States and timers are not supported for the moment.");
+        !signature.usesState() && !signature.usesTimers(),
+        "States and timers are not supported for the moment.");
 
     checkState(
-        !DoFnSignatures.requiresTimeSortedInput(doFn),
+        signature.onWindowExpiration() == null, "onWindowExpiration is not supported: %s", doFn);
+
+    checkState(
+        !signature.processElement().requiresTimeSortedInput(),
         "@RequiresTimeSortedInput is not supported for the moment");
 
+    SparkSideInputReader.validateMaterializations(transform.getSideInputs().values());
+    return true;
```

Comment:
for ParDos yes 👍  false is used to tell the translator to not translate a composite, but instead rather explode it further

Reply:
Yes that is why it should not return a boolean since it makes confusing. Instead, either passing by and return void or through an exception.

Reply:
Not sure I understand your points @aromanenko-dev... Have you missed the `@Override` ?
```
@Override
public boolean canTranslate(ParDo.MultiOutput<InputT, OutputT> transform) { ... }
```

This overrides a default implementation that has to support 3 different cases:
- `true` if the PTransform can be translated (that's the return value of the default implementation)
- `false` if a composite `PTransform` cannot be translated and should be further expanded into it's parts (see [this example](https://github.com/apache/beam/blob/master/runners/spark/3/src/main/java/org/apache/beam/runners/spark/structuredstreaming/translation/batch/CombineGroupedValuesTranslatorBatch.java#L58-L60) in case of `CombineWithContext`)
- throw an exception if usage of unsupported features is detected

Reply:
Yes, I missed `@Override` there with this git diff representation since I thought it's related to `translate()` method .
Then it's ok though still a bit confusing since it's not clear that it throws an exception without looking into the code.

Reply:
The right place to check in that case is the javadocs of the default implementation I suppose. Should I rephrase / clarify the javadocs further or does it make sense as is?
```
  /**
   * Checks if a composite / primitive transform can be translated. Composites that cannot be
   * translated as is, will be exploded further for translation of their parts.
   *
   * <p>This should be overridden where necessary. If a transform is know to be unsupported, this
   * should throw a runtime exception to give early feedback before any part of the pipeline is run.
   */
  protected boolean canTranslate(TransformT transform) 
  ```

Reply:
No, I think it's fine since it explains this in Javadoc.

Should not be `RuntimeException` added to method signature?

Reply:
👍 updated